src/proc/analyzer.rs

```diff
@@ -0,0 +1,517 @@
+        /// This flag bubbles up from child expressions to parents.
+        const REQUIRE_UNIFORM = 0x2;
+        /// The code may exit the control flow.
+        ///
+        /// `Kill` and `Return` operations have an effect on all the other
+        /// expressions/statemetns that are evaluated after them. They act as
+        /// pervasive control flow branching. Thus, even after the remaining
+        /// merge together, we are blocked from consideing the merged flow uniform.
+        const MAY_EXIT = 0x4;
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct SamplingKey {
+    pub image: Handle<crate::GlobalVariable>,
+    pub sampler: Handle<crate::GlobalVariable>,
+}
+
+#[derive(Clone, Default)]
+pub struct ExpressionInfo {
+    pub control_flags: ControlFlags,
+    pub ref_count: usize,
+}
+
+pub struct FunctionInfo {
+    pub control_flags: ControlFlags,
+    pub sampling_set: crate::FastHashSet<SamplingKey>,
+    expressions: Box<[ExpressionInfo]>,
+}
+
+impl ops::Index<Handle<crate::Expression>> for FunctionInfo {
+    type Output = ExpressionInfo;
+    fn index(&self, handle: Handle<crate::Expression>) -> &ExpressionInfo {
+        &self.expressions[handle.index()]
+    }
+}
+
+#[derive(Clone, Debug, thiserror::Error)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum AnalysisError {
+    #[error("Expression {0:?} is not a global variable!")]
+    ExpectedGlobalVariable(crate::Expression),
+    //TODO: add more information here!
+    #[error("Required uniformity of control flow is not fulfilled")]
+    NonUniformControlFlow,
+}
+
+impl FunctionInfo {
+    #[must_use]
+    fn add_ref(&mut self, handle: Handle<crate::Expression>) -> ControlFlags {
+        let info = &mut self.expressions[handle.index()];
+        info.ref_count += 1;
+        info.control_flags
+    }
+
+    /// Computes the control flags of a given expression, and store them
+    /// in `self.expressions`. Also, bumps the reference counts on
+    /// dependent expressions.
+    fn process_expression(
+        &mut self,
+        handle: Handle<crate::Expression>,
+        expression_arena: &Arena<crate::Expression>,
+        global_var_arena: &Arena<crate::GlobalVariable>,
+        other_functions: &[FunctionInfo],
+    ) -> Result<(), AnalysisError> {
+        use crate::{Expression as E, SampleLevel as Sl};
+
+        let control_flags = match expression_arena[handle] {
+            E::Access { base, index } => self.add_ref(base) | self.add_ref(index),
+            E::AccessIndex { base, .. } => self.add_ref(base),
+            E::Constant(_) => ControlFlags::empty(),
+            E::Compose { ref components, .. } => {
+                let mut accum = ControlFlags::empty();
+                for &comp in components {
+                    accum |= self.add_ref(comp);
+                }
+                accum
+            }
+            E::FunctionArgument(_) => ControlFlags::NON_UNIFORM_RESULT, //TODO?
+            E::GlobalVariable(handle) => {
```

Comment:
Is there a reference we could use to clarify the rules about which can be uniform? Some of them seem fairly intuitive, but others (like `Sc::Storage`) aren't obvious to me

Reply:
I don't know the source doc for this, but I'll sprinkle more comments.